Mathlib/Analysis/Calculus/FDeriv/Analytic.lean

```diff
@@ -294,3 +294,85 @@ theorem CPolynomialOn.iterated_deriv (h : CPolynomialOn 𝕜 f s) (n : ℕ) :
+    simp only [toFormalMultilinearSeries, h.symm, dite_false]
+
+open Finset in
+theorem changeOrigin_toFormalMultilinearSeries [DecidableEq ι] (x : ∀ i, M₁ i) :
+    continuousMultilinearCurryFin1 R (∀ i, M₁ i) M₂ (f.toFormalMultilinearSeries.changeOrigin x 1) =
+    f.linearDeriv x := by
+  ext y
+  rw [continuousMultilinearCurryFin1_apply, linearDeriv_apply,
+      changeOrigin, FormalMultilinearSeries.sum]
+  cases he : isEmpty_or_nonempty ι
+  · have (l) : 1 + l ≠ Fintype.card ι := by
+      rw [add_comm, Fintype.card_eq_zero]; exact Nat.succ_ne_zero _
+    simp_rw [Fintype.sum_empty, changeOriginSeries_support _ (this _), zero_apply _, tsum_zero]; rfl
+  rw [tsum_eq_single (Fintype.card ι - 1), changeOriginSeries]; swap
+  · intro m hm
+    rw [Ne, eq_tsub_iff_add_eq_of_le (by exact Fintype.card_pos), add_comm] at hm
+    rw [f.changeOriginSeries_support hm, zero_apply]
+  have heq : Fin.snoc 0 y = fun _ : Fin 1 ↦ y := by
+    ext; rw [Fin.snoc, dif_neg (Nat.not_lt_zero _)]; rfl
+  rw [sum_apply, ContinuousMultilinearMap.sum_apply, heq]
+  simp_rw [changeOriginSeriesTerm_apply]
+  refine (Fintype.sum_bijective (?_ ∘ Fintype.equivFinOfCardEq (Nat.add_sub_of_le
+    Fintype.card_pos).symm) (.comp ?_ <| Equiv.bijective _) _ _ fun i ↦ ?_).symm
+  · exact (⟨{·}ᶜ, by
+      rw [card_compl, Fintype.card_fin, card_singleton, Nat.add_sub_cancel_left]⟩)
+  · use fun _ _ ↦ (singleton_injective <| compl_injective <| Subtype.ext_iff.mp ·)
+    intro ⟨s, hs⟩
+    have h : sᶜ.card = 1 := by rw [card_compl, hs, Fintype.card_fin, Nat.add_sub_cancel]
+    obtain ⟨a, ha⟩ := card_eq_one.mp h
+    refine ⟨a, Subtype.ext (compl_eq_comm.mp ha)⟩
+  rw [Function.comp_apply, Subtype.coe_mk, compl_singleton, piecewise_erase_univ,
+    toFormalMultilinearSeries, dif_pos (Nat.add_sub_of_le Fintype.card_pos).symm]
+  simp_rw [domDomCongr_apply, compContinuousLinearMap_apply, ContinuousLinearMap.proj_apply,
+    Function.update_apply, (Equiv.injective _).eq_iff, ite_apply]
+  congr; ext j
+  obtain rfl | hj := eq_or_ne j i
+  · rw [Function.update_same, if_pos rfl]
+  · rw [Function.update_noteq hj, if_neg hj]
+
+protected theorem hasFDerivAt [DecidableEq ι] (x : ∀ i, M₁ i) :
+    HasFDerivAt f (f.linearDeriv x) x := by
+  rw [← changeOrigin_toFormalMultilinearSeries]
+  convert f.hasFiniteFPowerSeriesOnBall.hasFDerivAt (y := x) ENNReal.coe_lt_top
+  rw [zero_add]
+
+variable (f : ContinuousMultilinearMap R M₁ M₂) {n : ℕ∞} (x : (i : ι) → M₁ i)
+
+lemma cPolynomialAt : CPolynomialAt R f x :=
+  f.hasFiniteFPowerSeriesOnBall.cPolynomialAt_of_mem
+    (by simp only [Metric.emetric_ball_top, Set.mem_univ])
+
+lemma cPolyomialOn : CPolynomialOn R f ⊤ := fun x _ ↦ f.cPolynomialAt x
+
+lemma contDiffAt : ContDiffAt R n f x := CPolynomialAt.contDiffAt (f.cPolynomialAt x)
+
+lemma contDiffOn (s : Set ((i : ι) → M₁ i)) : ContDiffOn R n f s :=
+  CPolynomialOn.contDiffOn (fun x _ ↦ f.cPolynomialAt x)
```

Comment:
Why use an arbitrary set `s` rather than `⊤` in `contDiffOn`?
```suggestion
lemma cPolyomialOn : CPolynomialOn R f ⊤ := fun x _ ↦ f.cPolynomialAt x

lemma contDiffAt : ContDiffAt R n f x := CPolynomialAt.contDiffAt (f.cPolynomialAt x)

lemma contDiffOn (s : Set ((i : ι) → M₁ i)) : ContDiffOn R n f s :=
  CPolynomialOn.contDiffOn (fun x _ ↦ f.cPolynomialAt x)
```

Reply:
In fact think that `ContinuousMultilinearMap.contDiffOn` can just be deleted, as you can easily get it by applying `ContDiff.contDiffOn` to `ContinuousMultilinearMap.contDiff`.

Reply:
Okay, deleting this
```suggestion
lemma cPolyomialOn : CPolynomialOn R f ⊤ := fun x _ ↦ f.cPolynomialAt x

lemma contDiffAt : ContDiffAt R n f x := CPolynomialAt.contDiffAt (f.cPolynomialAt x)
```